vocab/cvocab.cc

```diff
@@ -194,17 +194,27 @@ void Vocab::group_ngrams(std::vector<std::string> & tokens,
     for (std::vector<std::string>::iterator it = tokens.begin();
          it != tokens.end(); ++it)
     {
+        // nothing to do if token was combined with another
+        if (*it == "")
```

Comment:
Instead of comparing against the `""` empty string literal. It would be better to check emptiness directly:
```c++
if (it->empty())
    continue;
```
The current form constructs an implicit temporary `std::string` object in order to perform the equality check. The given form avoids creating a temporary entirely and can do a more efficient check.